### PR TITLE
Clip renaming context menu item

### PR DESCRIPTION
**Pull request by Paul-Licameli**
Resolves: #1660

This makes #998 and #1614 more complete

Add an item to the wave clip context menu for renaming.

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

File: src/tracks/playabletrack/wavetrack/ui/WaveTrackView.cpp [hunk 1]
```diff
@@ -824,6 +826,24 @@ auto WaveTrackSubView::GetMenuItems(
...
+      if (pTrack) {
+         auto &view = TrackView::Get(*pTrack);
+         auto pAffordance = view.GetAffordanceControls();
+         if ( auto affordanceControl =
+            std::dynamic_pointer_cast<WaveTrackAffordanceControls>(
+               pAffordance ) ) {
+            // Begin editing of the label, either with a dialog or directly
+            // in the track panel.
+            auto &project = context.project;
+            affordanceControl->StartEditSelectedClipName( project,
```
Comment (vsverchinsky) [inline src/tracks/playabletrack/wavetrack/ui/WaveTrackView.cpp:839]: Suggest adding `WaveTrackAffordanceControls::StartEditClipNameAt(AudacityProject&, double time)`  instead, or use events

Comment (Paul-Licameli) [reply src/tracks/playabletrack/wavetrack/ui/WaveTrackView.cpp:839]: I don't understand the suggestion to use events.

The new test argument makes the function very general, but maybe you think it is too general.  If you don't, you might object that the generalized function should at least be renamed, and I would agree.

Comment (vsverchinsky) [reply src/tracks/playabletrack/wavetrack/ui/WaveTrackView.cpp:839]: It's not about generality. Yes, function name now became at bit controversial.

Comment (vsverchinsky) [reply src/tracks/playabletrack/wavetrack/ui/WaveTrackView.cpp:839]: Also, right channel does not always have affordance cell, may be in that case call should be redirected to the view of the leader track?

Comment (Paul-Licameli) [reply src/tracks/playabletrack/wavetrack/ui/WaveTrackView.cpp:839]: > Also, right channel does not always have affordance cell, may be in that case call should be redirected to the view of the leader track?

That is a good point.

Comment (Paul-Licameli) [reply src/tracks/playabletrack/wavetrack/ui/WaveTrackView.cpp:839]: If events are used, which object would emit the event and which would react to it?

**Comment by Paul-Licameli** [comment]
I have changed the function name and also check other channels.

I still do not understand why event notifications should be used here.  You used them so that changed of track selection state could be detected by TextEditHelper, and that was proper for avoidance of dependency cycles (TrackList should not depend on the code that reacts to the event).  I don't see a problem with cycles in this commit that needs this technique to break them.

**Comment by Paul-Licameli** [comment]
I also added the neglected change of undo history.

**Comment by Paul-Licameli** [comment]
The problem of pushing the undo history stack isn't solved, so I mark this as a draft until I do that.

**Comment by Paul-Licameli** [comment]
I don't recall why I thought undo history was sometimes not pushed -- maybe I was confused.

But it can be unnecessarily pushed -- if you just hit Enter with no changes to the text.  I changed that case.  It should affect behavior of clicking on the name as well as the new context menu item.

Esc key works as you would expect, discarding any changes you may have made without hitting Enter.

**Review by vsverchinsky — APPROVED** [review]

**Pull request merged**